blackjax/adaptation/adjusted_mclmc_adaptation.py

```diff
@@ -0,0 +1,378 @@
+    dual_averaging_adaptation,
+)
+from blackjax.diagnostics import effective_sample_size
+from blackjax.util import incremental_value_update, pytree_size
+
+Lratio_lowerbound = 0.0
+Lratio_upperbound = 2.0
+
+
+def adjusted_mclmc_find_L_and_step_size(
+    mclmc_kernel,
+    num_steps,
+    state,
+    rng_key,
+    target,
+    frac_tune1=0.1,
+    frac_tune2=0.1,
+    frac_tune3=0.1,
+    diagonal_preconditioning=True,
+    params=None,
+    max="avg",
+    num_windows=1,
+    tuning_factor=1.3,
+):
+    """
+    Finds the optimal value of the parameters for the MH-MCHMC algorithm.
+
+    Parameters
+    ----------
+    mclmc_kernel
+        The kernel function used for the MCMC algorithm.
+    num_steps
+        The number of MCMC steps that will subsequently be run, after tuning.
+    state
+        The initial state of the MCMC algorithm.
+    rng_key
+        The random number generator key.
+    target
+        The target acceptance rate for the step size adaptation.
+    frac_tune1
+        The fraction of tuning for the first step of the adaptation.
+    frac_tune2
+        The fraction of tuning for the second step of the adaptation.
+    frac_tune3
+        The fraction of tuning for the third step of the adaptation.
+    diagonal_preconditioning
+        Whether to do diagonal preconditioning (i.e. a mass matrix)
+    params
+        Initial params to start tuning from (optional)
+    max
+        whether to calculate L from maximum or average eigenvalue. Average is advised.
+    num_windows
+        how many iterations of the tuning are carried out
+    tuning_factor
+        multiplicative factor for L
+
+
+    Returns
+    -------
+    A tuple containing the final state of the MCMC algorithm and the final hyperparameters.
+    """
+
+    frac_tune1 /= num_windows
+    frac_tune2 /= num_windows
+    frac_tune3 /= num_windows
+
+    dim = pytree_size(state.position)
+    if params is None:
+        params = MCLMCAdaptationState(
+            jnp.sqrt(dim), jnp.sqrt(dim) * 0.2, sqrt_diag_cov=jnp.ones((dim,))
+        )
+
+    part1_key, part2_key = jax.random.split(rng_key, 2)
+
+    for i in range(num_windows):
+        window_key = jax.random.fold_in(part1_key, i)
+        (state, params, eigenvector) = adjusted_mclmc_make_L_step_size_adaptation(
+            kernel=mclmc_kernel,
+            dim=dim,
+            frac_tune1=frac_tune1,
+            frac_tune2=frac_tune2,
+            target=target,
+            diagonal_preconditioning=diagonal_preconditioning,
+            max=max,
+            tuning_factor=tuning_factor,
+        )(state, params, num_steps, window_key)
+
+    if frac_tune3 != 0:
+        for i in range(num_windows):
+            part2_key = jax.random.fold_in(part2_key, i)
+            part2_key1, part2_key2 = jax.random.split(part2_key, 2)
+
+            state, params = adjusted_mclmc_make_adaptation_L(
+                mclmc_kernel,
+                frac=frac_tune3,
+                Lfactor=0.5,
```

Comment:
I dont quite understand some of the parameter choice, but i trust that you have test them or they are mostly arbitory.

Reply:
Actually this is a good reminder: I updated the tuning so this part is off by default, since it doesn't in general yield an improvement - this is reflected in the sampling book section too. (I want it to be available though, so have not removed it from the code)